nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py

```diff
@@ -0,0 +1,255 @@
+                    input_ids=input_ids,
+                    position_ids=position_ids,
+                    encoder_input=None,
+                    attention_mask=attention_mask,
+                    labels=labels,
+                    set_inference_key_value_memory=set_inference_key_value_memory,
+                    inference_max_sequence_len=inference_max_sequence_len,
+                )
+
+        return output
+
+    def setup(self, stage=None):
+        if stage == 'predict' or self.virtual_prompt_style == VirtualPromptStyle.INFERENCE:
+            self.frozen_model.freeze()
+            return
+
+        self.setup_test_data()
+        if stage == 'test':
+            return
+
+        self.setup_training_data()
+        self.setup_validation_data()
+        logging.info(f'setup completed:\n{self.frozen_model.summarize()}')
+
+    def on_train_end(self):
+        # Save the best nemo model
+        self.save_to(save_path=self.cfg.nemo_path)
+
+    def get_forward_output_only_func(self):
+        """
+        Used for generate method only for now.
+        """
+
+        def fwd_output_only_func(batch, model):
+            extra_arg = {}
+            (
+                tokens,
+                attention_mask,
+                position_ids,
+                task_ids,
+                set_inference_key_value_memory,
+                inference_max_sequence_len,
+            ) = batch
+
+            tokens = tokens.cuda()
+            attention_mask = attention_mask.cuda()
+            position_ids = position_ids.cuda()
+            task_ids = task_ids.cuda()
+            extra_arg['set_inference_key_value_memory'] = set_inference_key_value_memory[0].item()
+            extra_arg['inference_max_sequence_len'] = inference_max_sequence_len[0].item()
+
+            output_tensor = model(tokens, position_ids, attention_mask, task_ids, **extra_arg)
+
+            def id_func(output_tensor):
+                return output_tensor, {'logits': output_tensor}
+
+            return output_tensor, id_func
+
+        return fwd_output_only_func
+
+    def state_dict(self, destination=None, prefix=None, keep_vars=False):
+        """
+        Creates a state_dict using only the adapter parameters.
+        This ensures that this wrapper class will only checkpoint the adapter
+        weights and not the rest of the base GPT Model.
+        """
+        state_dict_ = {}
+        for name, module in self.frozen_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    adapter_module = module.adapter_layer[adapter_key]
+                    state_adapter_key = ':'.join([name, adapter_key])
+                    state_dict_[state_adapter_key] = adapter_module.state_dict()
+
+                module.set_enabled_adapters(enabled=True)
+        return state_dict_
+
+    def load_state_dict(self, state_dict, strict: bool = True):
```

Comment:
Have you tried running this with `megatron_amp_O2`? It injects an extra `.module` that may mess with state dict save/load.

Reply:
No I haven't tested with O2. Prompt learning also converts O2 to non-O2 so I'm following that pattern.